File: C++/sum-of-root-to-leaf-binary-numbers.cpp
// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};


class Solution {
public:
    int sumRootToLeaf(TreeNode* root) {
        // Time:  O(n)
        // Space: O(h)
        return sumNode(root, 0);
    }
private:
    int sumNode(TreeNode* node, int val) {
        if (!node) {
            return 0;
        }

        val = val * 2 + node->val;
        if (!node->left && !node->right) {
            return val;
        }

        return sumNode(node->left, val) + sumNode(node->right, val);
    }
};

File: Python/sum-of-root-to-leaf-binary-numbers.py
from typing import Optional


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def sumRootToLeaf(self, root: Optional[TreeNode]) -> int:
        # Time:  O(n)
        # Space: O(h)
        return self.sum_node(root, 0)

    def sum_node(self, node, val):
        if not node:
            return 0

        val = 2 * val + node.val

        if not (node.left or node.right):
            return val

        return self.sum_node(node.left, val) + self.sum_node(node.right, val)
